nixos/luksroot: add tryEmptyPassphrase option

therishidesai opened the pull request:
###### Description of changes
Added an option that will try an empty passphrase to unlock the disk if the keyFile failed or wasn't given.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
    - Added two new NixOS test's:
       - [With systemd in initrd](https://github.com/therishidesai/nixpkgs/blob/bb34a519a84af4bbd97081994395b5026d953a74/nixos/tests/systemd-initrd-luks-empty-passphrase.nix)
       - [Without systemd in initrd](https://github.com/therishidesai/nixpkgs/blob/bb34a519a84af4bbd97081994395b5026d953a74/nixos/tests/initrd-luks-empty-passphrase.nix)
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

therishidesai commented:
Main motivation for this change is the support the `crypttab` `try-empty-password` option and have a similar option for systems that don't use systemd and crypttab. Specifically, we want the feature described in the crypttab manpage:

```
Takes a boolean argument. If enabled, right before asking the user for a password it is first attempted to unlock the volume with an empty password. This is useful for systems that are initialized with an encrypted volume with only an empty password set, which shall be replaced with a suitable password during first boot, but after activation.
```

therishidesai commented:
CC: @peterwilli @ck3d

ck3d reviewed (approved):
The code and test looks good to me. I am interested in how people will use this option in a secure way.

peterwilli commented:
Hey there! I'm not in the position to approve or deny anything, but from what I have read in the source I see nothing that rings any bells in my head.

So if I get this right, this function is for initialized, encrypted volumes with no password, but can be set later. Is that for OEMs to sell encrypted PC's? I'm just curious.

Also, I assume this can be used for encryption with key files alone as well, right?

therishidesai commented:
Yeah the main reason for this feature is for devices that have an encrypted disk but allow for setting the passphrase later.

therishidesai commented:
@dasJ looks like your review is blocking this

ck3d commented on nixos/modules/system/boot/luksroot.nix:
```diff
@@ -207,19 +207,20 @@ let
                 fi
             done
-        ''}
+        '' else '' ''}
```
Is there a reason adding a space in the else case?

ck3d reviewed (approved):
LGTM, but please cleanup the commits.

RaitoBezarius commented:
Can you fix the conflicts and we can merge this?

therishidesai commented:
> Can you fix the conflicts and we can merge this?

resolved the conflicts

RaitoBezarius commented:
> > Can you fix the conflicts and we can merge this?
> 
> resolved the conflicts

There is a merge commit lurking, can you squash everything please? Thanks! Also, could we have a mention in notable changes for the release 23.05 for this option? Thank you!

therishidesai commented:
@RaitoBezarius Updated the release notes doc and squashed the hanging merge commit

peterwilli commented:
Is it finally happening?! Finally, we merge?! I'd love to help! Do you need help with resolving merge coflicts?

RaitoBezarius commented:
Thank you, I am a bit busy right now and I don't want to rush the merge button, so I will take some days to review this, if you don't hear me by next week, please spam me!

ElvishJerricco commented on nixos/modules/system/boot/luksroot.nix:
```diff
@@ -594,6 +625,26 @@ in
+            '';
+          };
+
+          keyFileTimeout = mkOption {
```
Does this option need to be implemented in scripted stage 1? Or at least given documentation saying it's only effective in systemd stage 1?

therishidesai replied:
I'm not sure why this argument will need to be in scripted stage 1. We already check if systemd is enabled with the following [assertion](https://github.com/NixOS/nixpkgs/blob/dedb14591a1e54e79d7fb92f4bcf253ef7e0c1b9/nixos/modules/system/boot/luksroot.nix#L946). And this actually get's used [here](https://github.com/NixOS/nixpkgs/blob/dedb14591a1e54e79d7fb92f4bcf253ef7e0c1b9/nixos/modules/system/boot/luksroot.nix#L515)

ElvishJerricco replied:
Actually, I think that assertion is wrong too. It says "When systemd stage 1 is enabled, all luks devices must have `dev.tryEmptyPassphrase` ***AND*** a non-null `dev.keyFileTimeout`," which would mean everyone always has to set it to true and have a timeout. So `dev.tryEmptyPassphrase && dev.keyFileTimeout != null` should be changed to `dev.tryEmptyPassphrase -> dev.keyFileTimeout != null` to say "All luks devices *that* have `dev.tryEmptyPassphrase` must have a non-null `dev.keyFileTimeout`."

Anyway, my point was that `keyFileTimeout` might as well be enabled for both systemd stage 1 and scripted stage 1, and it isn't implemented for scripted stage 1 here yet. But given that it's new functionality, it is not needed.

therishidesai replied:
Ah yeah good catch

ElvishJerricco commented on nixos/modules/system/boot/luksroot.nix:
```diff
@@ -207,19 +207,47 @@ let
+                      do_open_passphrase
+                 '' else "die"}
+            ''}
+        fi
```
I'm not entirely sure this block correctly handles the case where the user just wants to enter a password? `dev.fallbackToPassword` defaults to `false`, and in the `else` cases for those checks, you have `die` in each of them. Have you tested this PR with regular password entry?

Also, seems like a lot more echo'ing will happen basically no matter what, including messages about key files even when none is configured.

I haven't tested these questions myself, so let me know if I'm just following the code flow wrong.

therishidesai replied:
Looks like you are correct. I think changing `dev.fallbackToPassword` to default to `true` is the simplest fix and is the action most users will want. If a user doesn't want this then they can explicitly set this option to false

therishidesai replied:
@RaitoBezarius @ElvishJerricco @peterwilli Do you guys agree with this as the default behavior?

ElvishJerricco replied:
I do not agree with changing the default behavior. Neither possible default is inherently better than the other, so it should stay the same.

therishidesai replied:
I'm pretty sure what I described is what already happens: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/luksroot.nix#L221

ElvishJerricco replied:
@therishidesai That is in the `else` block. That only happens if `keyFile == null`. i.e. when `keyFile == null`, always use passphrase, but when `keyFile != null`, only fallback to passphrase if `fallbackToPassword == true`. This is the behavior we should maintain, and we should not change the default values of these things.

ElvishJerricco replied:
@therishidesai actually I feel like this entire code block is overengineered. I think this would have done the trick much more easily (untested)

```patch
diff --git a/nixos/modules/system/boot/luksroot.nix b/nixos/modules/system/boot/luksroot.nix
index bbdb26f975e..9ca26b44536 100644
--- a/nixos/modules/system/boot/luksroot.nix
+++ b/nixos/modules/system/boot/luksroot.nix
@@ -158,6 +158,13 @@ let
       wait_target "header" ${dev.header} || die "${dev.header} is unavailable"
     ''}
 
+    try_empty_passphrase() {
+        ${if dev.tryEmptyPassphrase then ''
+            echo " - trying empty passphrase..."
+            ${csopen} --key-file=/dev/null
+        '' else "return 1"}
+    }
+
     do_open_passphrase() {
         local passphrase
 
@@ -213,12 +220,16 @@ let
               ${optionalString (dev.keyFileSize != null) "--keyfile-size=${toString dev.keyFileSize}"} \
               ${optionalString (dev.keyFileOffset != null) "--keyfile-offset=${toString dev.keyFileOffset}"}
         else
-            ${if dev.fallbackToPassword then "echo" else "die"} "${dev.keyFile} is unavailable"
-            echo " - failing back to interactive password prompt"
-            do_open_passphrase
+            if ! try_empty_passphrase; then
+                ${if dev.fallbackToPassword then "echo" else "die"} "${dev.keyFile} is unavailable"
+                echo " - failing back to interactive password prompt"
+                do_open_passphrase
+            fi
         fi
         '' else ''
-        do_open_passphrase
+        if ! try_empty_passphrase; then
+            do_open_passphrase
+        fi
         ''}
     }
 

```

therishidesai commented:
@ElvishJerricco Finally got around to working on this, I cleaned it up in a similar way to what you suggested. One thing that wasn't clean was handling what to do if the key file was missing so I added the extra else block [here](https://github.com/NixOS/nixpkgs/pull/176828/files#diff-0e36ad76d4195902c9910dea27b7bcfcf390a5af1c811c920a11dea6db668429R240). I also fixed up the assertions as you recommended

ElvishJerricco commented on nixos/modules/system/boot/luksroot.nix:
```diff
@@ -212,13 +227,28 @@ let
             ${csopen} --key-file=${dev.keyFile} \
               ${optionalString (dev.keyFileSize != null) "--keyfile-size=${toString dev.keyFileSize}"} \
               ${optionalString (dev.keyFileOffset != null) "--keyfile-offset=${toString dev.keyFileOffset}"}
+            cs_status=$?
+            if [ $cs_status -ne 0 ]; then
+              echo "Key File ${dev.keyFile} failed!"
+              if ! try_empty_passphrase; then
+                ${if dev.fallbackToPassword then "echo" else "die"} "${dev.keyFile} is unavailable"
+                echo " - failing back to interactive password prompt"
+                do_open_passphrase
+              fi
+            fi
```
Why are you adding a password check when the key file fails? Before this change, `fallbackToPassword` only did anything if the key file was completely missing. If the file was present but wrong, the device failed.

therishidesai replied:
Added this for the case that you would want multiple ways of unlocking a disk

ElvishJerricco commented on nixos/modules/system/boot/luksroot.nix:
```diff
@@ -212,13 +227,28 @@ let
         fi
         '' else ''
-        do_open_passphrase
+           echo "No key file given"
```
No need for added verbosity

ElvishJerricco commented on nixos/modules/system/boot/luksroot.nix:
```diff
@@ -158,6 +158,21 @@ let
+    try_empty_passphrase() {
+        ${if dev.tryEmptyPassphrase then ''
+             echo "Trying empty passphrase!"
+             echo "" | ${csopen}
+             cs_status=$?
+             if [ $cs_status -eq 0 ]; then
+                 echo "Empty passphrase succeeded"
```
No need for the verbosity with "Empty passphrase succeeded". Plus you can just leave `${csopen}` as the last command, and that will cause its exit code to be returned.

ElvishJerricco commented on nixos/modules/system/boot/luksroot.nix:
```diff
@@ -889,6 +942,10 @@ in
           message = "boot.initrd.luks.devices.<name>.bypassWorkqueues is not supported for kernels older than 5.9";
         }
 
+        { assertion = all (dev: dev.tryEmptyPassphrase && dev.keyFileTimeout != null ) (attrValues luks.devices);
+          message = "boot.initrd.luks.devices.<name>.tryEmptyPassphrase requires a keyFileTimeout to be set";
+        }
+
```
This assertion is still wrong. It says that *all* devices *must* have `tryEmptyPassphrase` and `keyFileTimeout`. You want `->`, not `&&`.

ElvishJerricco replied:
Also is this assertion even true? `keyFileTimeout` isn't required with `tryEmptyPassphrase` with systemd stage 1, and it's not used at all in the regular initrd in this PR.

therishidesai replied:
yeah this is wrong, let me fix this

therishidesai replied:
Yeah I'm not sure why I had this in the first place. I don't think this is necessary.

ElvishJerricco reviewed (changes requested):
> One thing that wasn't clean was handling what to do if the key file was missing so I added the extra else block [here](https://github.com/NixOS/nixpkgs/pull/176828/files#diff-0e36ad76d4195902c9910dea27b7bcfcf390a5af1c811c920a11dea6db668429R240)

Hm? The else block was where the password fallback was already happening. You've *added* a new password fallback code path when the key file fails.

therishidesai commented:
> > One thing that wasn't clean was handling what to do if the key file was missing so I added the extra else block [here](https://github.com/NixOS/nixpkgs/pull/176828/files#diff-0e36ad76d4195902c9910dea27b7bcfcf390a5af1c811c920a11dea6db668429R240)
> 
> Hm? The else block was where the password fallback was already happening. You've _added_ a new password fallback code path when the key file fails.

So I removed the assertion since it wasn't necessary and cleaned up the extraneous commits. 

As far as the extra new password fallback after the keyfile fails, the main reason for adding this is to support multiple ways to unlock a key. There are various use cases where you may have old keyfiles that will fail but still want to get into a machine and have added a password to unlock the disk.

ElvishJerricco reviewed (changes requested):
Last couple of things that I see.

1. If you're only going to implement `keyFileTimeout` for systemd initrd and not the regular initrd, then there needs to be an assertion like:
    ```
    !config.boot.initrd.systemd.enable -> all (x: x.keyFileTimeout == null) (attrValues luks.devices)
    ```
    i.e. When systemd initrd is not enabled, all devices should not use the unimplemented feature.
2. I'd like to see the two test files unified. We should be testing the same use cases with systemd-initrd as regular initrd. For example, the `nixos/tests/hibernate.nix` file has a `systemdStage1 ? false` argument. And I think it'd be best to use the `bootctl set-default` method of boot entry selection in both cases. We're testing initrd, not grub vs systemd-boot :) Having the test just enter arrow keys seems fragile to me.

therishidesai commented:
@ElvishJerricco Updated the assertion and also removed updated the test to use bootctl and also it takes a `systemdStage1` arg so you can test systemd initrd using the `tryEmptyPassphrase`

ElvishJerricco commented:
@therishidesai Your nixos/tests/ files are not committed.

therishidesai commented:
> @therishidesai Your nixos/tests/ files are not committed.

fixed

ElvishJerricco commented on nixos/tests/all-tests.nix:
```diff
@@ -649,6 +650,7 @@ in {
   systemd-initrd-btrfs-raid = handleTest ./systemd-initrd-btrfs-raid.nix {};
   systemd-initrd-luks-fido2 = handleTest ./systemd-initrd-luks-fido2.nix {};
   systemd-initrd-luks-keyfile = handleTest ./systemd-initrd-luks-keyfile.nix {};
+  systemd-initrd-luks-empty-passphrase = handleTest ./systemd-initrd-luks-empty-passphrase.nix {};
```
This file doesn't exist. I'm guessing it's supposed to be like this?

```suggestion
  systemd-initrd-luks-empty-passphrase = handleTest ./initrd-luks-empty-passphrase.nix { systemdStage1 = true; };
```

therishidesai replied:
Yeah my mistake, will fix

ElvishJerricco commented on nixos/tests/initrd-luks-empty-passphrase.nix:
```diff
@@ -0,0 +1,97 @@
+    # Encrypt key with empty key so boot should try keyfile and then fallback to empty passphrase
+
+
+    def grub_select_boot_luks_wrong_key_file():
+        """
+        Selects "boot-luks" from the GRUB menu
+        to trigger a login request.
+        """
+        machine.send_monitor_command("sendkey down")
+        machine.send_monitor_command("sendkey down")
+        machine.send_monitor_command("sendkey ret")
+
+    def grub_select_boot_luks_missing_key_file():
+        """
+        Selects "boot-luks" from the GRUB menu
+        to trigger a login request.
+        """
+        machine.send_monitor_command("sendkey down")
+        machine.send_monitor_command("sendkey ret")
```
These `grub_select_boot_` functions can be deleted

ElvishJerricco commented on nixos/tests/initrd-luks-empty-passphrase.nix:
```diff
@@ -0,0 +1,97 @@
+          device = "/dev/vdc";
+          keyFile = "/etc/cryptroot.key";
+          tryEmptyPassphrase = true;
+          fallbackToPassword = !systemdStage1;
```
What's the reasoning for `fallbackToPassword = !systemdStage1` in the two specialised configs? The tests seem to pass without setting that.

therishidesai replied:
There is an assertion for systemd stage1 where fallbackToPassword is implied in systems stage 1

ElvishJerricco replied:
It fails if you *set it* to `true`, but if you just leave it out entirely it's fine and defaults to the right thing.

ElvishJerricco reviewed (changes requested):
Sorry for the long review process on this one! I think this should be the last round.

therishidesai commented:
@ElvishJerricco just bumping this again

therishidesai commented:
@ElvishJerricco Any chance you can put the final approval on this soon?

ElvishJerricco commented on nixos/modules/system/boot/luksroot.nix:
```diff
@@ -594,6 +625,26 @@ in
+            type = types.nullOr types.int;
+            description = ''
+              The amount of time in seconds for a keyFile to appear before
+              timing out and trying passwords. This is required if tryEmptyPassphrase
+              is enabled and systemd in initrd is enabled.
```
I don't think this description is correct. This is not required if `tryEmptyPassphrase` is enabled.

ElvishJerricco reviewed (changes requested):
Just a little documentation issue.

ElvishJerricco reviewed (changes requested):
One more little documentation issue. Not sure why that didn't show up in the github checks before.

therishidesai committed:
nixos/luksroot: add tryEmptyPassphrase option

ElvishJerricco commented:
@ofborg test luks initrd-luks-empty-passphrase systemd-initrd-luks-empty-passphrase

ElvishJerricco commented:
Hm the `luks` test is hanging on aarch64. But it seems like that's true on `nixos-unstable-small` as well, so it's unlikely to have anything to do with this PR. So I'll go ahead and merge, given that everything else passed and it looks good to me.
